KAFKA-12980: Return empty record batch from Consumer::poll when position advances due to aborted transactions

C0urante opened the pull request:
[Jira](https://issues.apache.org/jira/browse/KAFKA-12980)

This is useful for reads to the end of a topic that contains aborted or empty transactions. If an aborted transaction is at the end of the topic, the consumer can now be expected to return from `poll` if it advances past that aborted transaction, and users can query the consumer's latest `position` for the relevant topic partitions to see if it has managed to make it past the end of the topic (or rather, what was the end of the topic when the attempt to read to the end of the topic began).

For a concrete example of this logic, see the [`KafkaBasedLog::readToLogEnd`](https://github.com/apache/kafka/blob/5e5d5bff3bdaf807338ec9adeac982f8a5c98fbd/connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java#L322-L345) method that Connect employs to refresh its view of internal topics.

No new unit tests are added, but many existing ones are modified to ensure that aborted and empty transactions are detected and reported correctly by `Fetcher::collectFetch`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

C0urante commented:
@hachikuji could you take a look at this when you have a chance?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java:
```diff
@@ -0,0 +1,124 @@
+
+public class Fetch<K, V> {
+    private final Map<TopicPartition, List<ConsumerRecord<K, V>>> records;
+    private boolean containsAborts;
```
Could we clarify what this means (maybe with a better name)? From the implementation, it looks like we set this for either aborted data or markers. I also wanted to mention that it is also possible for the position to advance due to an empty committed transaction. It wouldn't be a common case, but nothing in the protocol forces a producer to write to each partition that is included in the transaction.

C0urante replied:
Based on your other comment I've altered this to `positionAdvanced`; LMKWYT.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1235,8 +1237,14 @@ public void assign(Collection<TopicPartition> partitions) {
+                final Fetch<K, V> fetch = pollForFetches(timer);
+                if (!fetch.isEmpty()) {
+                    if (fetch.records().isEmpty()) {
+                        assert fetch.containsAborts();
```
nit: we do not often use assertions. If you feel the check is worthwhile, maybe we can raise an IllegalStateException with a useful error?

C0urante replied:
I've removed the check; I think that the cost of throwing an error to the user is a little too high since the alternative (an empty batch of records) seems less harmful.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -638,15 +636,15 @@ public void onFailure(RuntimeException e) {
      * Return the fetched records, empty the record buffer and update the consumed position.
      *
-     * NOTE: returning empty records guarantees the consumed position are NOT updated.
+     * NOTE: returning an {@link Fetch#isEmpty empty} fetch guarantees the consumed position is not updated.
```
I was considering whether it would be possible to do something a little simpler without caring as much about the case of aborted data. In the future we may have more instances of control records which would probably need similar logic. The case of an empty committed transaction mentioned in the other comment is an example of this.

I think the main thing we're interested in is when the position advances (regardless of the reason). We want to ensure that we return from poll() whenever this happens. In the code, this happens here: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java#L744. Could we structure this logic so that `Fetch` is tracking all the positions which have advanced? Then the condition for returning would only consider positions which have advanced.

C0urante replied:
Thanks, this is much simpler. I've taken a stab at an approach that's agnostic with regard to the cause of the position advancing and only tracks whether any advancement has taken place; hope this looks something like what you had in mind.

C0urante committed:
KAFKA-12980: Return empty record batch from Consumer::poll when position advances

C0urante commented:
Thanks for the review @hachikuji; I agree that it's simpler and more future-proof to make the logic more general and not worry about the specific case of aborted transactions. I've updated the PR (and title) accordingly.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1235,9 +1236,15 @@ public void assign(Collection<TopicPartition> partitions) {
-                final Map<TopicPartition, List<ConsumerRecord<K, V>>> records = pollForFetches(timer);
-                if (!records.isEmpty()) {
-                    // before returning the fetched records, we can send off the next round of fetches
+                final Fetch<K, V> fetch = pollForFetches(timer);
```
Can we update the documentation for `poll()` to mention that it will return when the position advances due to aborted transactions or control records?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetch.java:
```diff
@@ -0,0 +1,123 @@
+            }
+        });
+    }
+}
```
nit: newline

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1235,9 +1236,15 @@ public void assign(Collection<TopicPartition> partitions) {
+                                        + "for at least one topic partition; this may happen in the case of aborted or empty transactions"
+                        );
+                        return ConsumerRecords.empty();
+                    }                    // before returning the fetched records, we can send off the next round of fetches
```
nit: move comment to following line

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1235,9 +1236,15 @@ public void assign(Collection<TopicPartition> partitions) {
+                final Fetch<K, V> fetch = pollForFetches(timer);
+                if (!fetch.isEmpty()) {
+                    if (fetch.records().isEmpty()) {
+                        log.debug(
```
I'm not so sure about the value of this log line. Maybe it would be more useful to ensure that we have enough logging in `Fetcher` that we can see when the position gets advanced without any record data? That would tell us not only why a poll() returned with empty data, but which partition caused it to do so. What do you think?

C0urante replied:
I think that works. My goal with this line was to help people understand the cause for this new behavior, as it may seem like a bug at first if `poll` returns empty batches before the poll timeout has elapsed. I've taken a stab at a message that has functional value (i.e., naming the specific topic partition whose position has advanced without any user-visible records) and still tries to let the user know about how/why this may affect `poll` behavior; LMKWYT.

Also, this section was pretty buggy as-was since it skipped the call to `client.transmitSends` and didn't pass the empty batch through the consumer's interceptors. I've fixed both of these; LMK if you think we should skip passing an empty batch to interceptors, though.

hachikuji reviewed (commented):
Looks pretty good overall. Left some small comments.

C0urante committed:
KAFKA-12980: Address review comments

C0urante commented:
Thanks @hachikuji. I've addressed the nits and tried to address the more substantial comment regarding logging; interested in your thoughts when you have a moment.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1176,9 +1176,10 @@ public void assign(Collection<TopicPartition> partitions) {
-     * This method returns immediately if there are records available. Otherwise, it will await the passed timeout.
-     * If the timeout expires, an empty record set will be returned. Note that this method may block beyond the
-     * timeout in order to execute custom {@link ConsumerRebalanceListener} callbacks.
+     * This method returns immediately if there are records available or if the position advances past control records.
```
nit: past control records or aborted transactions when isolation.level=READ_COMMITTED

C0urante replied:
Ack, will add.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -728,6 +728,15 @@ public void onFailure(RuntimeException e) {
+                    if (partRecords.isEmpty()) {
+                        log.debug(
+                                "Advanced position for partition {} without receiving any user-visible records. " 
+                                        + "This is likely due to skipping over control records in the current fetch, " 
```
How about this. First, we can augment the message above to something like this:
```scala
  log.trace("Updating fetch position from {} to {} for partition {} and returning {} records from `poll()`", 
    position, nextPosition, completedFetch.partition, partRecords.size());
```
This gives us enough information in the logs to see which partitions caused the `poll()` to return and it tell us exactly where in the log the aborted transaction/control records exist. Second, maybe we can add back your previous log line, but make it a little more terse and put it at trace level:
```scala
                        log.trace(
                                "Returning empty records from `poll()` since the consumer's position has advanced "
                                        + "for at least one topic partition")
```

C0urante replied:
Sounds good to me 👍

C0urante committed:
KAFKA-12980: Simplify new logging messages

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1176,7 +1176,8 @@ public void assign(Collection<TopicPartition> partitions) {
      * <p>
-     * This method returns immediately if there are records available or if the position advances past control records.
+     * This method returns immediately if there are records available or if the position advances past control records
+     * or aborted transactions when isolation.level=READ_COMMITTED.
```
nit: Worth double-checking, but I think we require "read_committed" to be lower case.

C0urante replied:
Sure, https://github.com/apache/kafka/blob/a2da309f134ec2930f70f1a3917eaade7548132f/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java#L557.

C0urante replied:
```suggestion
     * or aborted transactions when isolation.level=read_committed.
```

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -725,17 +725,13 @@ public void onFailure(RuntimeException e) {
-                                completedFetch.partition
-                        );
+                        log.trace("Returning empty records from `poll()` " 
+                                + "since the consumer's position has advanced for at least one topic partition");
```
nit: I think this comment made more sense in its original location in `KafkaConsumer`. At this level, it seems redundant after the changes in the log message above. We would already say "... and returning 0 records from `poll()`"

C0urante replied:
This is true. I was hoping we could have something in here that explicitly states that this can happen because of the change in behavior implemented in this PR (i.e., skipping control records or aborted transactions). If you think it's worth it to call that out in a log message we can do that here, otherwise the entire `if (partRecords.isEmpty())` branch is unnecessary.

hachikuji replied:
I'm inclined to either remove the log line entirely or move it back to its former location in `KafkaConsumer`. Will leave it up to you.

C0urante replied:
Moved it to `KafkaConsumer`.

hachikuji reviewed (approved):
LGTM overall. Just two minor comments.

C0urante committed:
Update clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

Co-authored-by: Jason Gustafson <jason@confluent.io>

C0urante committed:
KAFKA-12980: Move log message from Fetcher to KafkaConsumer

hachikuji reviewed (approved):
LGTM. Thanks for the updates.

nbali commented:
@C0urante FYI this has a functionality loss. If you `KafkaConsumer.poll(Duration)` you are unable to tell anymore if it returns empty if it was a timeout - so most likely the topic is fully consumed - or not. Right now even a single empty (numRecords = 0, positionAdvanced = true) fetch will trigger a return. Deduplicated KTables can return fetches like that.

C0urante commented:
@nbali this PR was merged over two years ago and I've lost almost all of the context around it. If you are seeing problems because of the changes made here, please file a Jira ticket describing what's going wrong (including examples and/or a practical use case if possible) and link it to the ticket for this change.

nbali commented:
@C0urante https://issues.apache.org/jira/browse/KAFKA-14865
